morl_baselines/multi_policy/multi_policy_moqlearning/mp_mo_q_learning.py

```diff
@@ -107,48 +106,144 @@ def get_config(self) -> dict:
+            timesteps_per_iteration: The number of timesteps per iteration.
+            eval_freq: The frequency of evaluation.
+            eval_env: The environment to use for evaluation.
+            ref_point: The reference point for the hypervolume calculation.
+            test_weights: The weight vectors to use for evaluation (e.g. for expected utility).
+            epsilon_linear_support: The epsilon value for the linear support algorithm.
+            num_episodes_eval: The number of episodes used to evaluate the value of a policy.
+        """
+        if eval_env is None:
+            eval_env = deepcopy(self.env)
+
+        if test_weights is None:
+            test_weights = equally_spaced_weights(self.reward_dim, n=64)
+
+        for iter in range(num_iterations):
+            if self.weight_selection_algo == "ols" or self.weight_selection_algo == "gpi-ls":
+                w = self.linear_support.next_weight(
+                    algo=self.weight_selection_algo,
+                    gpi_agent=self if self.weight_selection_algo == "gpi-ls" else None,
+                    env=eval_env if self.weight_selection_algo == "gpi-ls" else None,
+                    rep_eval=num_episodes_eval,
                 )
-            self.global_step += len(self.agents) * self.eval_freq
-            rewards, disc_rewards = self.eval_all_agents()
-            hv = hypervolume(self.ref_point, rewards)
+            elif self.weight_selection_algo == "random":
+                w = random_weights(self.reward_dim)
+
+            new_agent = MOQLearning(
+                env=self.env,
+                id=iter,
+                weights=w,
+                scalarization=self.scalarization,
+                learning_rate=self.learning_rate,
+                gamma=self.gamma,
+                initial_epsilon=self.initial_epsilon,
+                final_epsilon=self.final_epsilon,
+                epsilon_decay_steps=self.epsilon_decay_steps,
+                dyna=self.dyna,
+                dyna_updates=self.dyna_updates,
+                log=self.log,
+                parent_writer=self.writer,
+            )
+            if self.reuse_q_table and len(self.policies) > 0:
+                reuse_ind = np.argmax([np.dot(w, v) for v in self.linear_support.ccs])
+                new_agent.q_table = deepcopy(self.policies[reuse_ind].q_table)
+            self.policies.append(new_agent)
+
+            start_time = time.time()
+            new_agent.global_step = self.global_step
+            new_agent.train(
+                start_time=start_time,
+                total_timesteps=timesteps_per_iteration,
+                reset_num_timesteps=False,
+                eval_freq=eval_freq,
+                eval_env=eval_env,
+            )
+            self.global_step = new_agent.global_step
+
+            value = policy_evaluation_mo(agent=new_agent, env=eval_env, w=w, rep=num_episodes_eval)
+            removed_inds = self.linear_support.add_solution(value, w)
```

Comment:
So this means MPMOQL is limited to weighted sum in practice?

Reply:
For now, yes. Maybe I can make this part be only executed when algo is ols or gpi-ls, then it should work with random weights and other utilities